File: python-algorithm/ds_5_tree/tree_3_bst_property/L98-m.cpp
//
// Created by LiHeng on 2020/5/20.
//

//二叉查找树（BST）：根节点大于等于左子树所有节点，小于等于右子树所有节点。
//二叉查找树中序遍历有序。

#include <iostream>
#include <vector>
using  namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

//方法1 利用了二叉搜索树的中序遍历性质
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        vector<int> res;
        inorderTree(root, res);
        for(int i=1; i<res.size(); ++i)
            if(res[i] <= res[i-1])
                return false;
        return true;
    }

    void inorderTree(TreeNode* root, vector<int> & res){
        if (root == NULL)
            return;
        inorderTree(root->left, res);
        res.push_back(root->val);
        inorderTree(root->right, res);
    }
};

//仍然是中序遍历，中序遍历的栈写法,相当于只把左节点入栈了，即每次遍历到左节点入栈，然后直接访问右节点不入栈

#include <stack>
class Solution1 {
public:
    bool isValidBST(TreeNode* root) {
        stack<TreeNode*> stack;
        long long inorder = (long long)INT_MIN - 1;

        while (!stack.empty() || root != nullptr) {
            while (root != nullptr) {
                stack.push(root);
                root = root -> left;
            }
            root = stack.top();
            stack.pop();
            // 如果中序遍历得到的节点的值小于等于前一个 inorder，说明不是二叉搜索树
            if (root -> val <= inorder) return false;
            inorder = root -> val;
            root = root -> right;
        }
        return true;
    }
};


//方法3  递归写法
//判断每个节点和父节点的关系
class Solution2 {
public:
    bool helper(TreeNode* root, long long lower, long long upper) {
        if (root == nullptr) return true;
        if (root -> val <= lower || root -> val >= upper) return false;
        return helper(root -> left, lower, root -> val) && helper(root -> right, root -> val, upper);
    }
    bool isValidBST(TreeNode* root) {
        return helper(root, LONG_MIN, LONG_MAX);
    }
};

File: python-algorithm/ds_5_tree/tree_3_bst_property/L98-m.py
# -*- coding: utf-8 -*-


# 判断一棵树是否为二叉搜索树

# 递归
class Solution:
    def isValidBST(self, root):
        def BFS(root, left, right):
            if root is None:
                return True
            if left < root.val < right:
                return BFS(root.left, left, root.val) and BFS(root.right, root.val, right)
            else:
                return False
        return BFS(root, -float('inf'), float('inf'))


# 中序遍历二叉树；为了省空间，不采用全部记录下来的方式，设置一个变量记录上一个节点即可
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        if not root:
            return True
        stk = []
        pre = -float("inf")
        while stk != [] or root != None:
            while root:
                stk.append(root)
                root = root.left
            root = stk.pop(-1)
            if root.val <= pre:
                return False
            pre = root.val
            root = root.right
        return True